scripts/nagios/check-rabbitmq-queue

```diff
@@ -31,50 +36,147 @@ states = {
     3: "UNKNOWN"
 }
 
-pattern = re.compile(r'(\w+)\t(\d+)')
-output = subprocess.check_output(['/usr/sbin/rabbitmqctl', 'list_queues'], universal_newlines=True)
+MAX_SECONDS_TO_CLEAR_FOR_BURSTS = 600
+MAX_SECONDS_TO_CLEAR_NORMAL = 120
+CRITICAL_SECONDS_TO_CLEAR_FOR_BURSTS = 1200
+CRITICAL_SECONDS_TO_CLEAR_NORMAL = 240
+
+def analyze_queue_stats(queue_name: str, stats: Dict[str, Any]) -> Dict[str, Any]:
+    now = int(time.time())
+    if stats == {}:
+        return dict(status=3,
+                    name=queue_name,
+                    message='invalid or no stats data')
+
+    if now - stats['update_time'] > 300 and queue_counts_rabbitmqctl[queue_name] > 10:
+        # Queue isn't updating the stats file and has some events in the backlog, it's
+        # likely stuck.
+        return dict(status=2,
+                    name=queue_name,
+                    message='queue appears to be stuck, last update %s' % (stats['update_time'],))
+
+    current_size = stats['current_queue_size']
+    average_consume_time = stats['recent_average_consume_time']
+    if average_consume_time is None:
+        # Queue just started.
+        return dict(status=0,
+                    name=queue_name,
+                    message='')
+
+    expected_time_to_clear_backlog = current_size * average_consume_time
+    if now - stats['queue_last_emptied_timestamp'] > 300:
+        if expected_time_to_clear_backlog > MAX_SECONDS_TO_CLEAR_NORMAL:
+            status = 2 if expected_time_to_clear_backlog > CRITICAL_SECONDS_TO_CLEAR_NORMAL else 1
+            return dict(status=status,
+                        name=queue_name,
+                        message='clearing backlog will take too long: %ss, size: %s' % (
+                            expected_time_to_clear_backlog, current_size))
+    else:
+        # We slept recently, so treat this as a burst.
+        if expected_time_to_clear_backlog > MAX_SECONDS_TO_CLEAR_FOR_BURSTS:
+            status = 2 if expected_time_to_clear_backlog > CRITICAL_SECONDS_TO_CLEAR_FOR_BURSTS else 1
+            return dict(status=status,
+                        name=queue_name,
+                        message='clearing the burst will take too long: %ss, size: %s' % (
+                            expected_time_to_clear_backlog, current_size))
+
+    return dict(status=0,
+                name=queue_name,
+                message='')
+
+WARN_COUNT_THRESHOLD_DEFAULT = 10
+CRITICAL_COUNT_THRESHOLD_DEFAULT = 50
+def check_other_queues(queue_counts_dict: Dict[str, int]) -> List[Dict[str, Any]]:
+    """ Do a simple queue size check for queues whose workers don't publish stats files."""
```

Comment:
Which queue workers is this intended to cover?  I think it might just be the Tornado ones?

Reply:
```
message_sender
missedmessage_email_senders
notify_tornado
tornado_return
test_suite
```
are all the queues that show up in `rabbitmqctl list_queues` but don't have a worker in `queue_processors`.

Reply:
OK; all of those but `notify_tornado` are deleted from the code (and I guess we have no mechanism for removing them from rabbitmq).